File: src/main/java/com/emilianomenendez/veritran/BankAccount.java
package com.emilianomenendez.veritran;

public class BankAccount {
    private Customer owner;
    private Dollars balance;

    public static BankAccountBuilder newAccountOwnedBy(Customer accountOwner) {
        return new BankAccountBuilder(accountOwner);
    }

    public BankAccount(Customer owner, Dollars initialBalance) {
        this.owner = owner;
        this.balance = initialBalance;
    }

    public boolean isOwnedBy(Customer customer) {
        return this.owner == customer;
    }

    public Dollars getBalance() {
        return balance;
    }

    public void deposit(Dollars amountToDeposit) {
        balance = balance.plus(amountToDeposit);
    }

    public void withdraw(Dollars amountToWithdraw) throws InsufficientFundsException {
        assertSufficientFunds(amountToWithdraw);

        balance = balance.minus(amountToWithdraw);
    }

    public void transfer(BankAccount destinationAccount, Dollars amountToTransfer) throws InsufficientFundsException {
        BankTransfer.from(this)
                .to(destinationAccount)
                .transfer(amountToTransfer);
    }

    private void assertSufficientFunds(Dollars amountToWithdraw) throws InsufficientFundsException {
        if (amountToWithdraw.isGreaterThan(balance)) {
            throw new InsufficientFundsException();
        }
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/domain/BankAccount.kt
package ar.com.flow.bankaccount.domain

import ar.com.flow.Customer
import ar.com.flow.bankaccount.domain.transaction.receipt.Receipt
import ar.com.flow.bankaccount.ports.out.Statement
import ar.com.flow.money.Money

interface BankAccount {
    val owner: Customer
    val currency: Currency
    val initialBalance: Balance
    val balance: Balance
    val previousBalance: Balance
    fun deposit(amountToDeposit: Money)
    fun withdraw(amountToWithdraw: Money)
    fun transfer(amountToTransfer: Money, creditAccount: BankAccount)
    fun addReceipt(receipt: Receipt)
    val statement: Statement
    fun withdrawalLimitSupports(amount: Money): Boolean
}